### ISSUE TITLE
[BUG] Midisrv and usbmidi2 driver hang with out-of-sequence MT3 SysEx packets

### ISSUE THREAD

**Issue by sat-okada** [microsoft/MIDI#437]
**Describe the bug**
If I send a 1-byte length System Exclusive Start message (F0 xx) in MT3 format, MIDI Service gets stuck.
If this happens, MidiSrv cannot be stopped and restarting OS is required.

**To Reproduce**
1.Set the Roland UM-ONE mk2 to class-compliant mode and attach it to the UM-ONE to PC.
2.Driver uses UsbMidi2.
3.Start up two consoles, One is a monitor.
> midi ep monitor --verbose

Send on the other
> midi ep send 0x30110100 0x00000000

4.On the monitor window, press escape to stop monitoring.
Monitoring cannot be terminated.
And finally, it is necessary to restart OS to recover.

**Expected behavior**
Wait for the next SxsEx Continue message and get the first 1 byte to complete the SysEx Start data.

**Installer Name or Version**
・Windows.MIDI.Services.In-Box.Service.-.1.0.1-preview.7.24305.1438-x64.exe
・Windows.MIDI.Services.Tools.and.SDKs.1.0.1-preview.7.24305.1438-x64.exe
・USB MIDI 2.0 class driver (USBMIDI2_10.0.1.7.x64.zip)

**Desktop (please complete the following information):**
・OS: Windows 11 Pro Insider Preview Build 27729.rs_prerelease.241011-1428

**Device information, if this is with an external MIDI device:**
Roland UM-ONE mk2

**Application Information**
midi.exe console app.

**Additional context**
This issue is very similar to #300.

**Comment by MusicMaker** [comment]
Can't reproduce sending this data from device to host
https://github.com/user-attachments/assets/a590fdd7-c5d3-4769-8007-db60b44fd82b

**Comment by AmeNote-Michael** [comment]
As indicated by @MusicMaker - the command is invalid. With proper UMP Commands, the expected result is observed. Please confirm this issue can be closed @sat-okada. FYI @Psychlist1972 

Issuing command as a SYSEX COMPLETE with: _midi ep send 0x30010100 0_, gives expected result

 ![Image](https://github.com/user-attachments/assets/25cc3b0e-5ea2-41b1-88ca-eafe2988421f)

Alternatively you can issue the command as per this Issue, but then submit a zero sized sysex end message and the end result is the same:
_midi ep send 0x30110100 0
midi ep send 0x30300000 0_

![Image](https://github.com/user-attachments/assets/890df121-d630-458a-81bf-e3eb269b9f7f)

**Comment by Psychlist1972** [comment]
@AmeNote-Michael do you know where the hang is happening? Is it in the driver?

Is there an opportunity for the code to recover gracefully instead of DOS'ing all of MIDI?

**Comment by AmeNote-Michael** [comment]
@Psychlist1972 issue is in USB MIDI 1.0 there is no partial SYSEX data packet without completing SYSEX send. So until I receive remainder of a 3 byte SYSEX to send in USB MIDI 1.0 or the end of SYSEX, I will not send the packet.

If other MIDI data other than real time is received then the SYSEX is terminated with no further data sent - like is done in USB MIDI 1.0.

Note that the driver is not hung or blocking - it is just waiting for the next UMP data to be provided.

**Comment by AmeNote-Michael** [comment]
@Psychlist1972 looking at sequence of original post, there may be an issue which we need to discuss as group. When the MIDI console escapes and the driver is to be closed, I do not think the service closes the driver - so the internal driver state machines are not cleared - which could mean we get into a bad state in the driver that is not recoverable.

Need to think this through some. @garydan42 perhaps if you have some time we could get online and chat about this. Maybe there is an event I am missing, needs to be added or managed in service?

**Comment by Psychlist1972** [comment]
> [@Psychlist1972](https://github.com/Psychlist1972) looking at sequence of original post, there may be an issue which we need to discuss as group. When the MIDI console escapes and the driver is to be closed, I do not think the service closes the driver - so the internal driver state machines are not cleared - which could mean we get into a bad state in the driver that is not recoverable.
> 
> Need to think this through some. [@garydan42](https://github.com/garydan42) perhaps if you have some time we could get online and chat about this. Maybe there is an event I am missing, needs to be added or managed in service?

- Given that the driver is the usb MIDI 2 driver, this will be in the KS transport. I wanted to ensure it wasn't something in the KSA 
- Given that it is a MIDI 1.0 byte-format endpoint, there will be no long-lived discovery connection, so that's not the problem. But you should be aware that if there's a MIDI 2.0 device attached, the driver is never closed / cleaned-up because we need to maintain a connection to capture function block and other metadata updates.

I'll defer to @garydan42 as to what event to check for when closing the driver connection.

**Comment by AmeNote-MikeKent** [comment]
It is not clear how to resolve this if there is incomplete or stalled sysex data. Driver is waiting for a completed sysex message for USB MIDI 1.0 translation. If the  device or application sends further data, then the completed message  will be translated and sent. If another message, other than sysex or system realtime, is received, then the driver will terminate the sysex by sending the data received so far. 
We do not see any better solution. @Psychlist1972, we propose this issue should be closed.

**Comment by Psychlist1972** [comment]
> It is not clear how to resolve this if there is incomplete or stalled sysex data. Driver is waiting for a completed sysex message for USB MIDI 1.0 translation. If the device or application sends further data, then the completed message will be translated and sent. If another message, other than sysex or system realtime, is received, then the driver will terminate the sysex by sending the data received so far. We do not see any better solution. [@Psychlist1972](https://github.com/Psychlist1972), we propose this issue should be closed.

I need Gary to weigh in on this one. Having an incomplete message in the USB MIDI 2 driver cause the whole service to hang is not a good outcome, but I do not know if there's a service-side solution or if the driver needs to have some sort of timeout handling with sysex. I'm not sure where that fix needs to be. @AmeNote-Michael can you bring this up in our internal discussion?

**Comment by AmeNote-Michael** [comment]
@Psychlist1972 will do. This may be a combination to be handled by both service and driver.

We should ask, how is it handled for example by the Network connection? Same thing can occur there and in fact is more likely to occur.

**Comment by Psychlist1972** [comment]
> [@Psychlist1972](https://github.com/Psychlist1972) will do. This may be a combination to be handled by both service and driver.
> 
> Was thinking same issue can occur with Network driver, but actually maybe not. The USB driver is unique in that it does USB MIDI 1.0 and USB MIDI 2.0 translations.
> 
> Will discuss with [@garydan42](https://github.com/garydan42), how is this handled for the service in the conversion? Maybe same issue can happen there.

Correct. The network transport deals with UMP only, so it's not waiting for anything.

**Comment by Psychlist1972** [comment]
Tracking internally as well with bug # 56179219

**Comment by AmeNote-Michael** [comment]
@sat-okada can you please confirm with the in box driver (see issue #499 to ensure you are loading correct driver).

I have been testing and I cannot reproduce issue with the in box Canary build 27788.rs_prerelease.250131-1609.

The driver is correctly handling the message, queueing internally to driver until more SYSEX data provided to complete a USB MIDI1.0 SYSEX command or a SYSEX END provided. Also will terminate message if a non sysex command such as note on occurs with current sysex data stored being dropped.

**Comment by AmeNote-Michael** [comment]
@sat-okada thank you for confirming #499. Can you please also confirm is this one still an issue with latest Canary?

**Comment by sat-okada** [comment]
I confirmed with the latest version.
This issue was not reproduced.
